azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +101,242 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
+
+            extension_signature_policy = _PolicyEngine.__parse_bool(individual_policy, "signatureRequired", extension_name)
+            if extension_signature_policy is None:
+                extension_signature_policy = self._policy["extensionPolicies"]["signatureRequired"]
+
+            policy_to_add = {
+                "signatureRequired": extension_signature_policy
+            }
+            parsed_extensions_dict[extension_name] = policy_to_add
+
+        # Convert "extensions" to a case-folded dict for case-insensitive lookup
+        case_folded_extensions_dict = _CaseFoldedDict.from_dict(parsed_extensions_dict)
+        self._policy["extensionPolicies"]["extensions"] = case_folded_extensions_dict
+
+    @staticmethod
+    def __validate_schema(policy, schema):
+        """
+        Validate the provided policy against the schema - we only do a shallow check (no recursion into nested dicts).
+        If there is an unrecognized attribute, raise an error.
+        """
+        for key in policy:
+            if key not in schema:
+                raise InvalidPolicyError("attribute '{0}' is not defined in the policy schema. Please refer to the policy documentation "
+                                         "and change or remove this attribute accordingly.".format(key))
+
+    @staticmethod
+    def __parse_bool(policy, key, parent_attribute_name=None):
+        """
+        Raise error if policy[key] is not a boolean. Return none if the key does not exist.
+        For attributes that are used in multiple places (ex: "signatureRequired"), we specify the parent attribute in
+        the error message to avoid ambiguity.
+        """
+        value = policy.get(key)
+        if value is None:
+            return None
+
+        if not isinstance(value, bool):
+            if parent_attribute_name is None:
+                msg = ("invalid type {0} for attribute '{1}', please change to bool."
+                       .format(type(value).__name__, key))
+            else:
+                msg = ("invalid type {0} for attribute '{1}' in section '{2}', please change to bool."
+                       .format(type(value).__name__, key, parent_attribute_name))
+
+            raise InvalidPolicyError(msg)
+
+        return value
+
+    @staticmethod
+    def __parse_dict(policy, key):
+        value = policy.get(key)
+        if value is None:
+            return None
+
+        if not isinstance(value, dict):
+            raise InvalidPolicyError("invalid type '{0}' for attribute '{1}', please change to object.".format(type(value).__name__, key))
+
+        return value
+
+    @staticmethod
+    def __parse_string(policy, key):
+        value = policy.get(key)
+        if value is None:
+            return None
+
+        if not isinstance(value, (str, ustr)):
+            raise InvalidPolicyError("invalid type '{0}' for attribute '{1}', please change to object.".format(type(value).__name__, key))
```

Comment:
>  please change to object

should this be str?

Reply:
Yes, thanks. 

I was thinking about merging __parse_bool, __parse_dict, __parse_string into a single function, so we have a common error message between all of them. I ended up keeping them separate because I thought it made the code more readable but let me know if you have a preference.

Reply:
I think it would make sense to merge them. 
I also think the type validation can be done in __validate_schema by using the value of the attribute in _POLICY_SCHEMA as the expected type